src/App.tsx

```diff
@@ -1,24 +1,288 @@
+        });
+      })
+      .catch(() => {
+        setErrorMessage(ERROR_MESSAGES.unableToDeleteTodo);
+        throw new Error(ERROR_MESSAGES.unableToDeleteTodo);
+      })
+      .finally(() => {
+        setIsProsessingTodoIds((prevState) => {
+          return prevState.filter(id => id !== todoId);
+        });
+
+        textInputRef.current?.removeAttribute('disabled');
+        if (textInputRef.current) {
+          textInputRef.current.focus();
+        }
+      });
+  };
+
+  const handleUpdateTodo = (todo: Todo, newTodoTitle: string) => {
+    setErrorMessage('');
+    setIsProsessingTodoIds((prevState) => {
+      return [...prevState, todo.id];
+    });
+
+    return todoService.updateTodo({
+      id: todo.id,
+      title: newTodoTitle,
+      userId: todo.userId,
+      completed: todo.completed,
+    })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(oneTodo => {
+          if (oneTodo.id !== updatedTodo.id) {
+            return oneTodo;
+          }
+
+          return updatedTodo;
+        }));
+      })
+      .catch(() => {
+        setErrorMessage(ERROR_MESSAGES.unableToUpdateTodo);
+        throw new Error(ERROR_MESSAGES.unableToUpdateTodo);
+      })
+      .finally(() => {
+        setIsProsessingTodoIds((prevState) => {
+          return prevState.filter(id => id !== todo.id);
+        });
+      });
+  };
+
+  const preparedTodos = useMemo(() => {
+    return getFilteredTodos(todos, selectFilter);
+  }, [todos, selectFilter]);
+
+  const handleClearCopmpletedTodo = async () => {
+    const todoToDelete = preparedTodos.filter(todo => todo.completed);
+    const deletePromises = todoToDelete.map(todo => handleDeleteTodo(todo.id));
+
+    try {
+      await Promise.all(deletePromises);
+      setTodos((prevState) => prevState
+        .filter((todo) => !todo.completed));
+    } catch (error) {
+      setErrorMessage(ERROR_MESSAGES.unableToDeleteTodo);
+    }
+  };
+
+  const handleChangeTodoStatus = (todo: Todo) => {
+    setIsProsessingTodoIds(prevState => [...prevState, todo.id]);
+
+    return todoService.updateTodo({
+      id: todo.id,
+      title: todo.title,
+      userId: todo.userId,
+      completed: !todo.completed,
+    })
+      .then(todoWithUpdateStatus => {
+        setTodos(prevState => {
+          return prevState.map(todoItem => {
+            if (todoItem.id !== todoWithUpdateStatus.id) {
+              return todoItem;
+            }
+
+            return todoWithUpdateStatus;
+          });
+        });
+      })
+      .catch(() => {
+        setErrorMessage(ERROR_MESSAGES.unableToUpdateTodo);
+      })
+      .finally(() => {
+        setIsProsessingTodoIds(prevState => {
+          return prevState.filter(id => id !== todo.id);
+        });
+      });
+  };
+
+  const handleToogleAllButton = async (allTodos: Todo[]) => {
+    const isAllTodosTrueOrFalse
+      = allTodos.every(todo => todo.completed)
+    || allTodos.every(todo => !todo.completed);
```

Comment:
It looks strange. You can just check if all completed.
In this case, toggle status for all todos, in another case `filter` them and toggle only uncompleted.

Also, set the loader for todos that are changing so the user case see them.